### DISCUSSION TITLE
Using `rulerfmt` to style the ruler

### DISCUSSION THREAD

**Discussion by joelim-work** [gokcehan/lf#1358] [General]
## Description

I had an idea to take the existing design of `statfmt` and apply it to styling the ruler as well, because the current `ruler` option doesn't have any real support for customizing colors.

So I wrote up an implementation for a new `rulerfmt` option (roughly written with limited testing), which works like below:

- Placeholders for internal data
  - `%a` is the pressed keys
  - `%p` is the progress of file operations
  - `%s` is the number of files that are selected or being designated for cut/copied
  - `%f` shows is a filter is currently being applied
  - `%i` is the position of the cursor
  - `%t` is the number of files in the current directory
  - `%d` is the amount of free disk space remaining
- Placeholders for options
  - `%{lf_<name>}` for the value of options (e.g. `%{lf_selmode}`)
  - `%{lf_user_<name>}` for the value of user-defined options (e.g. `%{lf_user_foobar}`)
- Formatting for optional fields
  - Use `|` to divide the ruler into sections
  - Any section containing a failed placeholder substitution (blank result) is omitted

## Implementation

Copy the following code into the `ui.drawStatLine` function (some other boilerplate code for adding a new option is required but not shown here):

```go
// var reRulerSub = regexp.MustCompile(`%[apsfitd]|%\{([^}])+\}`)
if gOpts.rulerfmt != "" {
    opts := getOptsMap()
    rulerfmt := strings.ReplaceAll(gOpts.rulerfmt, "|", "\x1f")
    rulerfmt = reRulerSub.ReplaceAllStringFunc(rulerfmt, func(s string) string {
        result := ""
        switch s {
        case "%a":
            result = acc
        case "%p":
            result = strings.Join(progress, "  ")
        case "%s":
            result = strings.Join(selection, "  ")
        case "%f":
            if len(dir.filter) != 0 {
                result = "\033[34;7m F \033[0m"
            }
        case "%i":
            result = fmt.Sprint(ind)
        case "%t":
            result = fmt.Sprint(tot)
        case "%d":
            result = diskFree(dir.path)
        default:
            s = strings.TrimSuffix(strings.TrimPrefix(s, "%{"), "}")
            if val, ok := opts[s]; ok {
                result = formatRulerOpt(s, val)
            }
        }
        if result == "" {
            result = "\x00"
        }
        return result
    })
    ruler := ""
    for _, section := range strings.Split(rulerfmt, "\x1f") {
        if !strings.Contains(section, "\x00") {
            ruler += section
        }
    }
    ui.msgWin.printRight(ui.screen, 0, st, ruler)
    return
}
```

## Example

`lfrc` configuration (adapted from https://github.com/gokcehan/lf/pull/1347#issuecomment-1650487545):

```
set rulerfmt "%p  |%s  |\033[35m%i\033[0m/\033[36m%t\033[0m|  \033[34m%d\033[0m|  \033[32m%{lf_user_selected_size}\033[0m"

cmd on-select &{{
    s=$(find $fx -type f -maxdepth 1 -ls | awk '{total += $7} END {print total}' | numfmt --to=iec --format='%.2fB')
	lf -remote "send $id set user_selected_size $s"
}}

map <space> :toggle; on-select; down
map v :invert; on-select
map u :unselect; on-select
```

![image](https://github.com/gokcehan/lf/assets/50560759/00338921-1c96-41e0-ae36-51ec9a80af39)

The value of `rulerfmt` can be a bit long, but it should be possible to build it in a `${{ ... }}` shell block which calls `lf -remote` at the end to set the value.

## Conclusion

This idea opens up more opportunities for customizing the ruler, especially the ability to apply styling and colors. I am not sure whether this will end up being merged, perhaps not since there is already a `ruler` option.

In any case, I didn't spend too long hacking this proof of concept up, as the main ideas were already implemented in the `statfmt` option, and the experience was fun anyway.

**Comment by gokcehan** [comment]
@joelim-work In general, I like the idea of introducing `rulerfmt` as it makes things more consistent with our options. `ruler` option is still relatively new (introduced in r29), so maybe a breaking change is still acceptable. If `rulerfmt` works in parallel with `ruler`, then we can also consider deprecating `ruler` for a few releases before removing it. So feel free to merge this if you like.

After `rulerfmt`, I think maybe the only inconsistency left would be to use the special separator `|` for `promptfmt` as well. I think maybe it could be introduced without any breaking changes. I'm not sure if there is a use case for it in `promptfmt` though.

**Reply by joelim-work** [reply]
@gokcehan Thanks for the feedback, I have submitted a PR #1386 which implements this option and deprecates the existing `ruler` option (it's still available for the time being).

I have thought about `promptfmt` too, but TBH I'm not sure if it's worth it because unlike the other cases there aren't really any 'optional' fields (you pretty much always have a username, hostname and current file). The only exception is the filter, but in my opinion this kind of information really belongs in the ruler, as you previously suggested in https://github.com/gokcehan/lf/pull/667#issuecomment-903816473.
